simulation/sub8_gazebo/diagnostics/job_runner.py

```diff
@@ -1,231 +1,118 @@
-                msg_time = self.cache_dict[key].get_last_message_time()
-                if msg is not None:
-                    if msg_time < self.nh.get_time() - rospy.Duration(.5):
-                        # If the message is too old, forget about it.
-                        continue
-
-                    if msg_time == last_timestamps[key]:
-                        continue
-                    last_timestamps[key] = msg_time
-
-                    msgs.append([key, (yield msg), (yield msg_time)])
-                else:
-                    print "There's a problem recording {0}".format(key)
-
-            yield self.nh.sleep(self.time_step)
-
-        print "Saving pre-fail data. Do not exit."
-        # We've written the post crash stuff now let's write the pre-crash data.
-        for i in range(self.buffer_array_size):
-            msgs = next(gen)
-            if msgs is not None:
-                for msg in msgs:
-                    bag.write(msg[0], msg[1], t=msg[2])
-        bag.close()
-        self.dumping = False
-
-    def write_to_cache(self, data):
-        self.cache[self.cache_index] = data
-
-        self.cache_index += 1
-        if self.cache_index >= self.buffer_array_size:
-            self.cache_index = 0
-
-    def read_from_cache(self):
-        index = self.cache_index
-        while True:
-            yield self.cache[index]
-            index += 1
-            if index >= self.buffer_array_size:
-                index = 0
-
-
 class JobManager(object):
-    def __init__(self, nh, loop_count=10000):
+    def __init__(self, nh, sub, log=True):
         self.nh = nh
+        self.sub = sub
         self.timedout = False
+        self.job_queue = deque()
+        self.record_log = log
 
-        self.bagger = BagManager(self.nh)
-        self.loop_count = loop_count
+        self.bagger = BagManager(self.nh, DIAG_OUT_DIR)
         self.successes = 0
         self.fails = 0
+        # TODO: append-to-queue service
 
-        self.run_job = nh.get_service_client('/gazebo/job_runner/mission_start', RunJob)
+    @txros.util.cancellableInlineCallbacks
+    def run(self):
+        """Let the JobManager free-run"""
+        while(len(self.job_queue) > 0):
+            yield self.run_next_job()
 
     @txros.util.cancellableInlineCallbacks
-    def run_job_loop(self):
-        print "Looking for jobs..."
+    def run_next_job(self):
+        """Run the next job in the job queue"""
+
+        job_constructor, loop_count = self.job_queue.popleft()
+
+        current_job = job_constructor(self.nh)
+
         current_loop = 0
 
-        self.bagger.start_caching()
+        if self.record_log:
+            self.bagger.start_caching()
+
+        while current_loop < loop_count:
+            current_loop += 1
+            try:
+                yield current_job.setup()
+            except Exception, error:
+                print "On test #{}, a {} raised an error on test setup".format(
```

Comment:
We could print everything by default and let the user pipe the output into a file only if they desire?